Fix race condition of deleting ccnp in e2e test

jschwinger233 opened the pull request:
There is a flake in e2e test when a test case starts to proceed before ccnp/cnp comes to take effect by cilium-agent. The correct way to delete ccnp/cnp is to run "kubectl delete" followed by "cilium policy wait", and kubectl helper already has such wrappers.

Fixes: #24380

```release-note
Fix race conditions when deleting CNP / CCNP in e2e tests
```

christarazi reviewed (approved):
Nice find!

julianwiedmann commented on test/k8s/datapath_configuration.go:
```diff
@@ -649,18 +649,24 @@ func fetchPodsWithOffset(kubectl *helpers.Kubectl, namespace, name, filter, host
 }
 
-func applyL3Policy(kubectl *helpers.Kubectl, ns string) {
+func applyL3Policy(kubectl *helpers.Kubectl, ns string) (withdrawPolicy func()) {
 	demoPolicyL3 := helpers.ManifestGet(kubectl.BasePath(), "l3-policy-demo.yaml")
 	By(fmt.Sprintf("Applying policy %s", demoPolicyL3))
 	_, err := kubectl.CiliumPolicyAction(ns, demoPolicyL3, helpers.KubectlApply, helpers.HelperTimeout)
 	ExpectWithOffset(1, err).Should(BeNil(), fmt.Sprintf("Error creating resource %s: %s", demoPolicyL3, err))
+
+	return func() {
+		_, err := kubectl.CiliumPolicyAction(ns, demoPolicyL3, helpers.KubectlDelete, helpers.HelperTimeout)
+		ExpectWithOffset(1, err).Should(BeNil(), fmt.Sprintf("Error deleting resource %s: %s", demoPolicyL3, err))
+	}
```
It's a bit confusing that the patch description speaks about fixing CCNP troubles, but here we also fix a CNP case. Could you split this part off into a second patch? The effective code delta will remain the same, so we can merge without re-running CI. 

Fixing it in the same PR is ok, in the end it's the same pattern of bug.

jschwinger233 replied:
Will do

pchaigno replied:
> so we can merge without re-running CI.

But don't rebase if you want to do that! :smile:

jschwinger233 committed:
test: fix race condition of deleting ccnp in e2e test

There is a flake in e2e test when a test case starts to proceed
before ccnp comes to take effect by cilium-agent. The correct way to
delete ccnp is to run "kubectl delete" followed by "cilium policy wait",
and kubectl helper already has such wrappers.

Fixes: #24380

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

jschwinger233 committed:
test: fix race condition of deleting cnp in e2e test

There is a flake in e2e test when a test case starts to proceed
before cnp comes to take effect by cilium-agent. The correct way to
delete cnp is to run "kubectl delete" followed by "cilium policy wait",
and kubectl helper already has such wrappers.

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>
